d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,576 @@
+  private final Map<String, ResourceSubscriber> _cdsResourceSubscribers = new HashMap<>();
+  private final Map<String, ResourceSubscriber> _edsResourceSubscribers = new HashMap<>();
+
+  private AdsStream _adsStream;
+
+  public XdsClientImpl(Node node, ManagedChannel managedChannel, ExecutorService executorService)
+  {
+    _node = node;
+    _managedChannel = managedChannel;
+    _executorService = executorService;
+  }
+
+  private void startRpcStream()
+  {
+    AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub =
+        AggregatedDiscoveryServiceGrpc.newStub(_managedChannel);
+    _adsStream = new AdsStream(stub);
+    _adsStream.start();
+    _log.info("ADS stream started");
+  }
+
+  @Override
+  void watchXdsResource(String resourceName, ResourceType type, ResourceWatcher watcher)
+  {
+    _executorService.execute(() ->
+    {
+      Map<String, ResourceSubscriber> resourceSubscriberMap = getResourceSubscriberMap(type);
+      ResourceSubscriber subscriber = resourceSubscriberMap.get(resourceName);
+      if (subscriber == null)
+      {
+        _log.info("Subscribe {} resource {}", type, resourceName);
+        subscriber = new ResourceSubscriber(type, resourceName);
+        resourceSubscriberMap.put(resourceName, subscriber);
+
+        if (_adsStream == null)
+        {
+          startRpcStream();
+        }
+        _adsStream.sendDiscoveryRequest(type, Collections.singletonList(resourceName));
+      }
+      subscriber.addWatcher(watcher);
+    });
+  }
+
+  private void handleLdsResponse(DiscoveryResponseData responseData)
+  {
+    Map<String, LdsUpdate> ldsUpdates = new HashMap<>();
+    List<String> errors = new ArrayList<>();
+
+    for (Resource resource : responseData.getResourcesList())
+    {
+      try
+      {
+        Listener listener = resource.getResource().unpack(Listener.class);
+
+        for (FilterChain filterChain : listener.getFilterChainsList())
+        {
+          for (Filter filter : filterChain.getFiltersList())
+          {
+            if (filter.hasTypedConfig() && filter.getTypedConfig()
+                .getTypeUrl()
+                .equals(TYPE_URL_HTTP_CONNECTION_MANAGER))
+            {
+              HttpConnectionManager httpConnectionManager = filter.getTypedConfig().unpack(HttpConnectionManager.class);
+              ldsUpdates.put(listener.getName(), new LdsUpdate(httpConnectionManager));
+            }
+          }
+        }
+      } catch (InvalidProtocolBufferException e)
+      {
+        _log.warn("Failed to unpack LDS response", e);
+        errors.add("Failed to unpack LDS response");
+      }
+    }
+
+    _log.info("Received LDS response nonce {}, resource updates {}", responseData.getNonce(), ldsUpdates);
+    handleResourceUpdate(ldsUpdates, responseData.getResourceType(), responseData.getNonce(), errors);
+  }
+
+  private void handleRdsResponse(DiscoveryResponseData responseData)
+  {
+    Map<String, RdsUpdate> rdsUpdates = new HashMap<>();
```

Comment:
If we move name directly into *dsUpdate, then here we can just use a Set. Having name as a field in the objects can help debugging at different levels. Also, two resources with the same content but different names should be considered as different resources by their Eq and Hash functions.